File: packages/py-parity-checking/sidecars/spiceypy/src/py_parity_sidecar/handlers/frames.py
from __future__ import annotations

from typing import Any

import spiceypy as sp
from spiceypy.utils.exceptions import NotFoundError
from spiceypy.utils.support_types import SPICEDOUBLE_CELL, SPICEINT_CELL

from ..models import (
    StepFramesCcifrm,
    StepFramesCidfrm,
    StepFramesCkcov,
    StepFramesCkgp,
    StepFramesCkgpav,
    StepFramesCklpf,
    StepFramesCkobj,
    StepFramesCkupf,
    StepFramesCnmfrm,
    StepFramesFrinfo,
    StepFramesFrmnam,
    StepFramesNamfrm,
    StepFramesPxform,
    StepFramesSxform,
    StepOutput,
    WorkflowStep,
)
from ..runtime import SidecarRuntimeContext, resolve_path_ref


def _to_matrix3(value: Any) -> list[list[float]]:
    rows_raw = value.tolist() if hasattr(value, "tolist") else value
    if not isinstance(rows_raw, list) or len(rows_raw) != 3:
        raise ValueError(f"Expected 3x3 matrix, got: {rows_raw!r}")

    out: list[list[float]] = []
    for row in rows_raw:
        if not isinstance(row, list) or len(row) != 3:
            raise ValueError(f"Expected 3x3 matrix row, got: {row!r}")
        out.append([float(row[0]), float(row[1]), float(row[2])])
    return out


def _to_matrix6(value: Any) -> list[list[float]]:
    rows_raw = value.tolist() if hasattr(value, "tolist") else value
    if not isinstance(rows_raw, list) or len(rows_raw) != 6:
        raise ValueError(f"Expected 6x6 matrix, got: {rows_raw!r}")

    out: list[list[float]] = []
    for row in rows_raw:
        if not isinstance(row, list) or len(row) != 6:
            raise ValueError(f"Expected 6x6 matrix row, got: {row!r}")
        out.append([float(v) for v in row])
    return out


def _to_vec3(value: Any) -> list[float]:
    values_raw = value.tolist() if hasattr(value, "tolist") else value
    if not isinstance(values_raw, list) or len(values_raw) != 3:
        raise ValueError(f"Expected length-3 vector, got: {values_raw!r}")
    return [float(values_raw[0]), float(values_raw[1]), float(values_raw[2])]


def _normalize_namfrm(value: Any) -> dict[str, Any]:
    if isinstance(value, tuple) and len(value) == 2 and isinstance(value[1], (bool, int)):
        code_raw, found_raw = value
        if not bool(found_raw):
            return {"found": False}
        code = int(code_raw)
        if code == 0:
            return {"found": False}
        return {"found": True, "code": code}

    code = int(value)
    if code == 0:
        return {"found": False}
    return {"found": True, "code": code}


def _normalize_frmnam(value: Any) -> dict[str, Any]:
    if isinstance(value, tuple) and len(value) == 2 and isinstance(value[1], (bool, int)):
        name_raw, found_raw = value
        if not bool(found_raw):
            return {"found": False}
        name = str(name_raw)
        if name.strip() == "":
            return {"found": False}
        return {"found": True, "name": name}

    name = str(value)
    if name.strip() == "":
        return {"found": False}
    return {"found": True, "name": name}


def _normalize_frcode_frname(value: Any) -> dict[str, Any]:
    if isinstance(value, tuple):
        if len(value) == 3 and isinstance(value[2], (bool, int)):
            frcode_raw, frname_raw, found_raw = value
            if not bool(found_raw):
                return {"found": False}
            return {
                "found": True,
                "frcode": int(frcode_raw),
                "frname": str(frname_raw),
            }
        if len(value) == 2:
            frcode_raw, frname_raw = value
            return {
                "found": True,
                "frcode": int(frcode_raw),
                "frname": str(frname_raw),
            }
    raise ValueError(f"Unexpected frame lookup shape: {value!r}")


def _normalize_frinfo(value: Any) -> dict[str, Any]:
    if isinstance(value, tuple):
        if len(value) == 4 and isinstance(value[3], (bool, int)):
            center_raw, frame_class_raw, class_id_raw, found_raw = value
            if not bool(found_raw):
                return {"found": False}
            return {
                "found": True,
                "center": int(center_raw),
                "frameClass": int(frame_class_raw),
                "classId": int(class_id_raw),
            }
        if len(value) == 3:
            center_raw, frame_class_raw, class_id_raw = value
            return {
                "found": True,
                "center": int(center_raw),
                "frameClass": int(frame_class_raw),
                "classId": int(class_id_raw),
            }
    raise ValueError(f"Unexpected frinfo shape: {value!r}")


def _normalize_ccifrm(value: Any) -> dict[str, Any]:
    if isinstance(value, tuple):
        if len(value) == 4 and isinstance(value[3], (bool, int)):
            frcode_raw, frname_raw, center_raw, found_raw = value
            if not bool(found_raw):
                return {"found": False}
            return {
                "found": True,
                "frcode": int(frcode_raw),
                "frname": str(frname_raw),
                "center": int(center_raw),
            }
        if len(value) == 3:
            frcode_raw, frname_raw, center_raw = value
            return {
                "found": True,
                "frcode": int(frcode_raw),
                "frname": str(frname_raw),
                "center": int(center_raw),
            }
    raise ValueError(f"Unexpected ccifrm shape: {value!r}")


def _normalize_ckgp(value: Any) -> dict[str, Any]:
    if isinstance(value, tuple):
        if len(value) == 3 and isinstance(value[2], (bool, int)):
            cmat_raw, clkout_raw, found_raw = value
            if not bool(found_raw):
                return {"found": False}
            return {
                "found": True,
                "cmat": _to_matrix3(cmat_raw),
                "clkout": float(clkout_raw),
            }
        if len(value) == 2:
            cmat_raw, clkout_raw = value
            return {
                "found": True,
                "cmat": _to_matrix3(cmat_raw),
                "clkout": float(clkout_raw),
            }
    raise ValueError(f"Unexpected ckgp shape: {value!r}")


def _normalize_ckgpav(value: Any) -> dict[str, Any]:
    if isinstance(value, tuple):
        if len(value) == 4 and isinstance(value[3], (bool, int)):
            cmat_raw, av_raw, clkout_raw, found_raw = value
            if not bool(found_raw):
                return {"found": False}
            return {
                "found": True,
                "cmat": _to_matrix3(cmat_raw),
                "av": _to_vec3(av_raw),
                "clkout": float(clkout_raw),
            }
        if len(value) == 3:
            cmat_raw, av_raw, clkout_raw = value
            return {
                "found": True,
                "cmat": _to_matrix3(cmat_raw),
                "av": _to_vec3(av_raw),
                "clkout": float(clkout_raw),
            }
    raise ValueError(f"Unexpected ckgpav shape: {value!r}")


def _set_ck_handle(context: SidecarRuntimeContext, handle_id: str, handle: int) -> None:
    if handle_id in context.state.frames.ckHandles:
        raise ValueError(f"CK handle already exists: {handle_id}")
    context.state.frames.ckHandles[handle_id] = int(handle)


def _require_ck_handle(context: SidecarRuntimeContext, handle_id: str) -> int:
    handle = context.state.frames.ckHandles.get(handle_id)
    if handle is None:
        raise ValueError(f"CK handle does not exist: {handle_id}")
    return handle


def _remove_ck_handle(context: SidecarRuntimeContext, handle_id: str) -> None:
    context.state.frames.ckHandles.pop(handle_id, None)


def _get_or_create_window(context: SidecarRuntimeContext, window_id: str, *, max_intervals: int) -> Any:
    existing = context.state.cellsWindows.windows.get(window_id)
    if existing is not None:
        return existing

    if max_intervals <= 0:
        raise ValueError("maxIntervals must be > 0")

    window = SPICEDOUBLE_CELL(max_intervals * 2)
    context.state.cellsWindows.windows[window_id] = window
    context.register_finalizer(
        f"cellsWindows.release:{window_id}",
        lambda: context.state.cellsWindows.windows.pop(window_id, None),
    )
    return window


def _window_to_intervals(window: Any) -> list[dict[str, float]]:
    n_intervals = int(sp.wncard(window))
    intervals: list[dict[str, float]] = []
    for index in range(n_intervals):
        left, right = sp.wnfetd(window, index)
        intervals.append({"left": float(left), "right": float(right)})
    return intervals


def run_frames_step(step: WorkflowStep, context: SidecarRuntimeContext) -> StepOutput | None:
    if isinstance(step, StepFramesNamfrm):
        out = _normalize_namfrm(sp.namfrm(step.name))
        return StepOutput(op=step.op, value=out)

    if isinstance(step, StepFramesFrmnam):
        out = _normalize_frmnam(sp.frmnam(step.code))
        return StepOutput(op=step.op, value=out)

    if isinstance(step, StepFramesCidfrm):
        try:
            out = _normalize_frcode_frname(sp.cidfrm(step.center))
            return StepOutput(op=step.op, value=out)
        except NotFoundError:
            return StepOutput(op=step.op, value={"found": False})

    if isinstance(step, StepFramesCnmfrm):
        try:
            out = _normalize_frcode_frname(sp.cnmfrm(step.centerName))
            return StepOutput(op=step.op, value=out)
        except NotFoundError:
            return StepOutput(op=step.op, value={"found": False})

    if isinstance(step, StepFramesFrinfo):
        try:
            out = _normalize_frinfo(sp.frinfo(step.frameId))
            return StepOutput(op=step.op, value=out)
        except NotFoundError:
            return StepOutput(op=step.op, value={"found": False})

    if isinstance(step, StepFramesCcifrm):
        try:
            out = _normalize_ccifrm(sp.ccifrm(step.frameClass, step.classId))
            return StepOutput(op=step.op, value=out)
        except NotFoundError:
            return StepOutput(op=step.op, value={"found": False})

    if isinstance(step, StepFramesCkgp):
        try:
            out = _normalize_ckgp(sp.ckgp(step.inst, step.sclkdp, step.tol, step.ref))
            return StepOutput(op=step.op, value=out)
        except NotFoundError:
            return StepOutput(op=step.op, value={"found": False})

    if isinstance(step, StepFramesCkgpav):
        try:
            out = _normalize_ckgpav(sp.ckgpav(step.inst, step.sclkdp, step.tol, step.ref))
            return StepOutput(op=step.op, value=out)
        except NotFoundError:
            return StepOutput(op=step.op, value={"found": False})

    if isinstance(step, StepFramesCklpf):
        ck_path = resolve_path_ref(context.paths, step.ck)
        handle = sp.cklpf(ck_path)
        _set_ck_handle(context, step.handleId, int(handle))
        return StepOutput(op=step.op, value={"opened": True})

    if isinstance(step, StepFramesCkupf):
        handle = _require_ck_handle(context, step.handleId)
        sp.ckupf(handle)
        _remove_ck_handle(context, step.handleId)
        return StepOutput(op=step.op, value={"closed": True})

    if isinstance(step, StepFramesCkobj):
        ck_path = resolve_path_ref(context.paths, step.ck)
        ids = SPICEINT_CELL(step.maxCard or 32)
        sp.scard(0, ids)
        sp.ckobj(ck_path, ids)
        return StepOutput(op=step.op, value={"ids": [int(v) for v in list(ids)]})

    if isinstance(step, StepFramesCkcov):
        ck_path = resolve_path_ref(context.paths, step.ck)
        cover = _get_or_create_window(
            context,
            step.coverId,
            max_intervals=step.maxIntervals or 128,
        )
        sp.scard(0, cover)
        sp.ckcov(ck_path, step.idcode, step.needav, step.level, step.tol, step.timsys, cover)
        return StepOutput(op=step.op, value={"intervals": _window_to_intervals(cover)})

    if isinstance(step, StepFramesPxform):
        out = sp.pxform(step.from_, step.to, step.et)
        return StepOutput(op=step.op, value=_to_matrix3(out))

    if isinstance(step, StepFramesSxform):
        out = sp.sxform(step.from_, step.to, step.et)
        return StepOutput(op=step.op, value=_to_matrix6(out))

    return None

File: packages/py-parity-checking/src/run-tspice/domains/frames.ts
import fs from "node:fs";

import type {
  Matrix6x6,
  PathRef,
  PathRefLike,
  StepFramesCcifrm,
  StepFramesCidfrm,
  StepFramesCkcov,
  StepFramesCkgp,
  StepFramesCkgpav,
  StepFramesCklpf,
  StepFramesCkobj,
  StepFramesCkupf,
  StepFramesCnmfrm,
  StepFramesFrinfo,
  StepFramesFrmnam,
  StepFramesNamfrm,
  StepFramesPxform,
  StepFramesSxform,
  StepOutput,
} from "../../case-types.js";
import { resolvePathRef, toPathRef, toVirtualKernelPath } from "../../fixtures.js";
import { getOrCreateWindow, type RunTspiceContext } from "../context.js";
import { unflattenMatrix } from "../utils/matrix.js";

type FramesStep =
  | StepFramesNamfrm
  | StepFramesFrmnam
  | StepFramesCidfrm
  | StepFramesCnmfrm
  | StepFramesFrinfo
  | StepFramesCcifrm
  | StepFramesCkgp
  | StepFramesCkgpav
  | StepFramesCklpf
  | StepFramesCkupf
  | StepFramesCkobj
  | StepFramesCkcov
  | StepFramesPxform
  | StepFramesSxform;

type FramesCkPathStep = StepFramesCklpf | StepFramesCkobj | StepFramesCkcov;

function pathRefsEqual(left: PathRef, right: PathRef): boolean {
  return left.kind === right.kind && left.rel === right.rel;
}

function canonicalizeFramesCkPath(context: RunTspiceContext, step: FramesCkPathStep): PathRef {
  const canonical = toPathRef(step.ck);

  for (const hint of context.normalization.metadata.runtimePath.canonicalizationHints) {
    if (hint.domain !== "frames" || hint.op !== step.op || hint.field !== "ck") {
      continue;
    }

    if (pathRefsEqual(hint.canonicalPath, canonical)) {
      return hint.canonicalPath;
    }
  }

  return canonical;
}

function ensureVirtualCkPath(context: RunTspiceContext, pathRef: PathRefLike): string {
  const virtualPath = toVirtualKernelPath(pathRef);
  if (!context.state.frames.stagedVirtualKernelPaths.has(virtualPath)) {
    const resolvedPath = resolvePathRef(context.paths, pathRef);
    const bytes = fs.readFileSync(resolvedPath);

    // Stage CK bytes into virtual FS once, then unload; cklpf controls loaded state thereafter.
    context.spice.raw.furnsh({ path: virtualPath, bytes });
    context.spice.raw.unload(virtualPath);
    context.state.frames.stagedVirtualKernelPaths.add(virtualPath);
  }

  return virtualPath;
}

function setCkHandle(context: RunTspiceContext, handleId: string, handle: number): void {
  if (context.state.frames.ckHandles.has(handleId)) {
    throw new Error(`CK handle already exists: ${handleId}`);
  }
  context.state.frames.ckHandles.set(handleId, handle);
}

function requireCkHandle(context: RunTspiceContext, handleId: string): number {
  const handle = context.state.frames.ckHandles.get(handleId);
  if (handle == null) {
    throw new Error(`CK handle does not exist: ${handleId}`);
  }
  return handle;
}

function deleteCkHandle(context: RunTspiceContext, handleId: string): void {
  context.state.frames.ckHandles.delete(handleId);
}

function requiredNumber(values: readonly number[], index: number): number {
  const value = values[index];
  if (value === undefined) {
    throw new Error(`Missing numeric value at index ${index}`);
  }
  return value;
}

function toVector3(values: readonly number[]): [number, number, number] {
  if (values.length !== 3) {
    throw new Error(`Expected length-3 vector, got ${values.length}`);
  }

  return [requiredNumber(values, 0), requiredNumber(values, 1), requiredNumber(values, 2)];
}

function toMatrix6x6(values: readonly number[]): Matrix6x6 {
  if (values.length !== 36) {
    throw new Error(`Expected 36 matrix values, got ${values.length}`);
  }

  const rows: Matrix6x6 = [
    [
      requiredNumber(values, 0),
      requiredNumber(values, 1),
      requiredNumber(values, 2),
      requiredNumber(values, 3),
      requiredNumber(values, 4),
      requiredNumber(values, 5),
    ],
    [
      requiredNumber(values, 6),
      requiredNumber(values, 7),
      requiredNumber(values, 8),
      requiredNumber(values, 9),
      requiredNumber(values, 10),
      requiredNumber(values, 11),
    ],
    [
      requiredNumber(values, 12),
      requiredNumber(values, 13),
      requiredNumber(values, 14),
      requiredNumber(values, 15),
      requiredNumber(values, 16),
      requiredNumber(values, 17),
    ],
    [
      requiredNumber(values, 18),
      requiredNumber(values, 19),
      requiredNumber(values, 20),
      requiredNumber(values, 21),
      requiredNumber(values, 22),
      requiredNumber(values, 23),
    ],
    [
      requiredNumber(values, 24),
      requiredNumber(values, 25),
      requiredNumber(values, 26),
      requiredNumber(values, 27),
      requiredNumber(values, 28),
      requiredNumber(values, 29),
    ],
    [
      requiredNumber(values, 30),
      requiredNumber(values, 31),
      requiredNumber(values, 32),
      requiredNumber(values, 33),
      requiredNumber(values, 34),
      requiredNumber(values, 35),
    ],
  ];
  return rows;
}

/** Execute one `frames.*` workflow step in tspice. */
export function runFramesStep(context: RunTspiceContext, step: FramesStep): StepOutput {
  switch (step.op) {
    case "frames.namfrm": {
      const out = context.spice.raw.namfrm(step.name);
      if (!out.found) {
        return { op: step.op, value: { found: false } };
      }
      return { op: step.op, value: { found: true, code: out.code } };
    }

    case "frames.frmnam": {
      const out = context.spice.raw.frmnam(step.code);
      if (!out.found) {
        return { op: step.op, value: { found: false } };
      }
      return { op: step.op, value: { found: true, name: out.name } };
    }

    case "frames.cidfrm": {
      const out = context.spice.raw.cidfrm(step.center);
      if (!out.found) {
        return { op: step.op, value: { found: false } };
      }
      return {
        op: step.op,
        value: {
          found: true,
          frcode: out.frcode,
          frname: out.frname,
        },
      };
    }

    case "frames.cnmfrm": {
      const out = context.spice.raw.cnmfrm(step.centerName);
      if (!out.found) {
        return { op: step.op, value: { found: false } };
      }
      return {
        op: step.op,
        value: {
          found: true,
          frcode: out.frcode,
          frname: out.frname,
        },
      };
    }

    case "frames.frinfo": {
      const out = context.spice.raw.frinfo(step.frameId);
      if (!out.found) {
        return { op: step.op, value: { found: false } };
      }
      return {
        op: step.op,
        value: {
          found: true,
          center: out.center,
          frameClass: out.frameClass,
          classId: out.classId,
        },
      };
    }

    case "frames.ccifrm": {
      const out = context.spice.raw.ccifrm(step.frameClass, step.classId);
      if (!out.found) {
        return { op: step.op, value: { found: false } };
      }
      return {
        op: step.op,
        value: {
          found: true,
          frcode: out.frcode,
          frname: out.frname,
          center: out.center,
        },
      };
    }

    case "frames.ckgp": {
      const out = context.spice.raw.ckgp(step.inst, step.sclkdp, step.tol, step.ref);
      if (!out.found) {
        return { op: step.op, value: { found: false } };
      }
      return {
        op: step.op,
        value: {
          found: true,
          cmat: unflattenMatrix(out.cmat),
          clkout: out.clkout,
        },
      };
    }

    case "frames.ckgpav": {
      const out = context.spice.raw.ckgpav(step.inst, step.sclkdp, step.tol, step.ref);
      if (!out.found) {
        return { op: step.op, value: { found: false } };
      }
      return {
        op: step.op,
        value: {
          found: true,
          cmat: unflattenMatrix(out.cmat),
          av: toVector3(out.av),
          clkout: out.clkout,
        },
      };
    }

    case "frames.cklpf": {
      const ck = ensureVirtualCkPath(context, canonicalizeFramesCkPath(context, step));
      const handle = context.spice.raw.cklpf(ck);
      setCkHandle(context, step.handleId, handle);
      return { op: step.op, value: { opened: true } };
    }

    case "frames.ckupf": {
      const handle = requireCkHandle(context, step.handleId);
      context.spice.raw.ckupf(handle);
      deleteCkHandle(context, step.handleId);
      return { op: step.op, value: { closed: true } };
    }

    case "frames.ckobj": {
      const ck = ensureVirtualCkPath(context, canonicalizeFramesCkPath(context, step));
      const ids = context.spice.kit.newIntCell(step.maxCard ?? 32);
      try {
        context.spice.raw.scard(0, ids);
        context.spice.raw.ckobj(ck, ids);

        const nIds = context.spice.raw.card(ids);
        const values: number[] = [];
        for (let index = 0; index < nIds; index += 1) {
          values.push(context.spice.kit.cellGeti(ids, index));
        }

        return {
          op: step.op,
          value: { ids: values },
        };
      } finally {
        context.spice.kit.freeCell(ids);
      }
    }

    case "frames.ckcov": {
      const ck = ensureVirtualCkPath(context, canonicalizeFramesCkPath(context, step));
      const cover = getOrCreateWindow(context, step.coverId, step.maxIntervals ?? 128);
      context.spice.raw.scard(0, cover);
      context.spice.raw.ckcov(
        ck,
        step.idcode,
        step.needav,
        step.level,
        step.tol,
        step.timsys,
        cover,
      );

      const nIntervals = context.spice.raw.wncard(cover);
      const intervals: Array<{ left: number; right: number }> = [];
      for (let index = 0; index < nIntervals; index += 1) {
        const [left, right] = context.spice.raw.wnfetd(cover, index);
        intervals.push({ left, right });
      }

      return {
        op: step.op,
        value: { intervals },
      };
    }

    case "frames.pxform": {
      const matrix = context.spice.raw.pxform(step.from, step.to, step.et);
      return { op: step.op, value: unflattenMatrix(matrix) };
    }

    case "frames.sxform": {
      const matrix = context.spice.raw.sxform(step.from, step.to, step.et);
      return { op: step.op, value: toMatrix6x6(matrix) };
    }

    default: {
      const exhaustive: never = step;
      throw new Error(`Unhandled frames step: ${JSON.stringify(exhaustive)}`);
    }
  }
}
